Deprecate Python 3.6 & 3.7; get CI working again

BrianPugh opened the pull request:
# Change Log
* Dropped python 3.6 and python 3.7 support.
* Added python 3.10 and python 3.11 to all the relevant places
* Updated pre-commit hooks to latest versions
* Replaced deprecated isort flag `-sl` with current `--force-single-line-imports`.
* Change "extras" dependencies to modern poetry dependency groups.
    * Removed all sphinx-related doc dependencies from the main dependency group.
* Fixed a few minor typing issues with `merge_styles`
* Explicitly set the `prompt_toolkit` version to be `<=3.0.36` because `3.0.37` has introduced some event-loop-related bug.
* `make clean` will now also clean the docs.


# Original Post
**What is the problem that this PR addresses?**
* isort 5.10.1 has the issues described [here](https://stackoverflow.com/questions/75269700/pre-commit-fails-to-install-isort-5-11-4-with-error-runtimeerror-the-poetry-co).

<!-- Describe the rationale of your PR. -->
<!-- Link all issues that it closes e.g. Closes #32 -->



**How did you solve it?**
* Ran `pre-commit autoupdate` to update all plugins; fixing minor linting issues in changes of Black.

**Checklist**
<!--- Put an `x` in all the boxes that apply. -->
<!-- Automated tests validate that your PR follows our guidelines. -->
<!-- If it doesn't pass, click on it to see details as to what your PR might be missing. -->

- [x] I have read the [Contributor's Guide](https://questionary.readthedocs.io/en/stable/pages/contributors.html#steps-for-submitting-code).
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

kiancross commented:
Thanks for the PR. It looks like the CI is broken for the repo at the moment. I will have to look into fixing this before I can merge!

BrianPugh commented:
Seems like issues with some CI running on older python 3.7. I'd recommend dropping support (it's at EOL in 2 months anyways)

kiancross commented:
@BrianPugh Could you submit a PR to drop support? I cannot approve my own PRs, so would need to wait until another maintainer is available to get it through...

BrianPugh commented:
on it 👍

kiancross commented:
Thanks, much appreciated. Would you be able to submit a separate PR for it?

BrianPugh commented:
I'm still working on fixing this up, I'll let you know when it's ready. Why the separate PR?

kiancross commented:
> I'm still working on fixing this up, I'll let you know when it's ready. Why the separate PR?

It's good practice to have only one change per PR. It also means the commit history has an entry per change. Finally, it means changes are not missed when the changelog is written.

kiancross commented:
Fair enough. Keep it in this one, and I'll make sure the commit description contains all the changes. Tag me when you're ready for me to review/merge :)

BrianPugh commented:
@kiancross I think this is getting close to complete. What I've done so far:
* Updated pre-commit hooks
* Replaced deprecated isort flag.
* Updated poetry
    * Updated dependency groups to the modern poetry >=1.2.0 format.
* Dropped python 3.6 and python 3.7 support.
* Added python 3.10 and python 3.11 to all the relevant places
* Fixed a few minor typing issues with `merge_styles`
* Explicitly set the `prompt_toolkit` version to be `<=3.0.36` because `3.0.37` has introduced some event-loop-related bug.
* `make clean` will now also clean the docs.

What I can't figure out:
* I've never really used intersphinx before, and I'm getting warnings (that get promoted to errors) like:

```
/Users/brianpugh/projects/questionary/questionary/question.py:docstring of questionary.question.Question:1: WARNING: py:class reference target not found: prompt_toolkit.application.application.Application
/Users/brianpugh/projects/questionary/questionary/prompts/text.py:docstring of questionary.prompts.text.text:33: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/text.py:docstring of questionary.prompts.text.text:41: WARNING: py:class reference target not found: prompt_toolkit.lexers.base.Lexer
/Users/brianpugh/projects/questionary/questionary/prompts/password.py:docstring of questionary.prompts.password.password:35: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/path.py:docstring of questionary.prompts.path.path:24: WARNING: py:class reference target not found: prompt_toolkit.shortcuts.prompt.CompleteStyle
/Users/brianpugh/projects/questionary/questionary/prompts/path.py:docstring of questionary.prompts.path.path:36: WARNING: py:class reference target not found: prompt_toolkit.completion.base.Completer
/Users/brianpugh/projects/questionary/questionary/prompts/path.py:docstring of questionary.prompts.path.path:39: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/confirm.py:docstring of questionary.prompts.confirm.confirm:28: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/select.py:docstring of questionary.prompts.select.select:47: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/rawselect.py:docstring of questionary.prompts.rawselect.rawselect:39: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/checkbox.py:docstring of questionary.prompts.checkbox.checkbox:51: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
/Users/brianpugh/projects/questionary/questionary/prompts/autocomplete.py:docstring of questionary.prompts.autocomplete.autocomplete:34: WARNING: py:class reference target not found: prompt_toolkit.completion.base.Completer
/Users/brianpugh/projects/questionary/questionary/prompts/autocomplete.py:docstring of questionary.prompts.autocomplete.autocomplete:45: WARNING: py:class reference target not found: prompt_toolkit.shortcuts.prompt.CompleteStyle
/Users/brianpugh/projects/questionary/questionary/prompts/autocomplete.py:docstring of questionary.prompts.autocomplete.autocomplete:57: WARNING: py:class reference target not found: prompt_toolkit.styles.style.Style
```

kiancross commented:
Thanks for all the help. If possible, try to keep the PR as limited in scope as possible (i.e., only what is needed to get CI working again).

> I've never really used intersphinx before, and I'm getting warnings (that get promoted to errors) like

Try changing the following URL

https://github.com/tmbo/questionary/blob/87a891c41a0199bb23b2b230e971fd15db32c17b/docs/conf.py#L36

to

https://python-prompt-toolkit.readthedocs.io/en/3.0.36/

If that doesn't work, I can take a look, but probs won't have time for a couple of weeks.

Another problem we have is that the 3.6 CI runs are marked as required, and I can't overrule this. @tmbo Can you remove all CI requirements from the branch rules for now, and I will provide a list of which ones to add back once all the CI issues are fixed?

BrianPugh commented:
> Thanks for all the help. If possible, try to keep the PR as limited in scope as possible (i.e., only what is needed to get CI working again).

All the changes so far are the minimum necessary to get CI working (outside of just disabling type checking).

> Try changing the following URL

That worked! looks like some codecov issues that i should be able to tack in the next day or two.

BrianPugh commented:
@kiancross so I *think* this all works, so I'd say approve the workflow to run and we'll see, but I think i tested it about as far as I can from my side.

kiancross commented:
The MacOS breakage might be because Poetry also needs updating (see #266).

BrianPugh commented:
updated; the cache might need to be cleared for it to take effect, though.

kiancross commented on .github/workflows/continuous-integration.yml:
```diff
@@ -55,113 +55,114 @@ jobs:
+      - name: Test Code 🔍
+        run: make test
+
+      - name: Send Coverage Report 📊
+        shell: bash -l {0}
+        env:
+          COVERALLS_REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          COVERALLS_SERVICE_NAME: github
+          COVERALLS_SERVICE_JOB_ID: ${{ github.run_id }}
+          COVERALLS_SERVICE_NUMBER: "${{ github.workflow }}-${{ github.run_number }}"
+          COVERALLS_PARALLEL: true
+        run: poetry run coveralls
```
Any idea why coveralls isn't posting a status on the PR anymore? Is it related to these changes? See this PR for an example of what used to happen https://github.com/tmbo/questionary/pull/244.

Also, source code is not visible in coveralls, but perhaps this was always the case e.g., https://coveralls.io/builds/60155724/source?filename=questionary%2Fprompts%2Fpath.py#L69. Is this related?

BrianPugh replied:
I don't have any experience with coveralls, this was actually a misguided change stemming from the discussion here:

https://github.com/TheKevJames/coveralls-python/issues/252

In reality, I just obviously didn't have permissions to the endpoint 😄 . We can remove these changes if you'd like.

kiancross replied:
Try removing and see if the status update comes back. Otherwise, I'll look at why the status has disappeared in a separate PR!

kiancross commented on .github/workflows/continuous-integration.yml:
```diff
@@ -55,113 +55,114 @@ jobs:
-    - name: Install dependencies 🖥
-      run: poetry install --no-interaction --extras "docs"
+      - name: Install dependencies 🖥
+        run: poetry install --no-interaction --with=docs
```
Is this just a syntactic change?

BrianPugh replied:
the spacing in this file is syntactic (I forgot to turn off my yaml linter).

Generally, `group` is used for internal stuff (like doc dependencies, dev dependencies, etc) with `with`, while `extras` is used for additional runtime features.

kiancross commented on docs/Makefile:
```diff
@@ -3,7 +3,7 @@
 # You can set these variables from the command line.
 SPHINXOPTS    = -W --keep-going -n
-SPHINXBUILD   = sphinx-build
+SPHINXBUILD   = python -m sphinx
```
Presumably no side-effects to this? Just a different way of invoking Sphinx?

BrianPugh replied:
this is just a more robust way of invoking sphinx-build; not sure when the change happened, but I (as well as CI) was having issues with poetry and sphinx getting along without this. See related issue:

https://github.com/sphinx-doc/sphinx/issues/11218

kiancross commented on pyproject.toml:
```diff
@@ -1,10 +1,10 @@
 [build-system]
-requires = ["poetry-core"]
+requires = ["poetry>=1.4.0"]
```
This was changed to `poetry-core` in #108. Do you know if one or the other is better?

BrianPugh replied:
I think `poetry-core` is actually more proper. The intent here was just that was had a modern version of poetry (at least 1.2.0).  I'll revert this (and apply it to some of my other projects 😄 )

kiancross commented on questionary/styles.py:
```diff
@@ -0,0 +1,12 @@
+from questionary.constants import DEFAULT_STYLE
+
+
+def merge_styles_default(styles: List[Optional[prompt_toolkit.styles.Style]]):
+    filtered_styles: list[prompt_toolkit.styles.BaseStyle] = [DEFAULT_STYLE]
+    filtered_styles.extend([s for s in styles if s is not None])
+    return prompt_toolkit.styles.merge_styles(filtered_styles)
```
What's the reasoning for this? Just to remove any `None` styles?

BrianPugh replied:
I believe prompt_toolkit has some improper type hinting. `merge_styles` can certainly handle `None` elements, but it's not in their type hints. So this fixes that so mypy passes.

kiancross replied:
Can you pop a comment in the source code explaining the reasoning?

kiancross reviewed (commented):
Ok, this is looking good! Thanks for all your help. I've left some comments/questions below. Can you also update the title of this PR to reflect it's current state and provide a list of changes in the description? I will use this for the commit when merging.

kiancross commented:
> Another problem we have is that the 3.6 CI runs are marked as required, and I can't overrule this.

This is sorted now too, so almost ready to merge!

BrianPugh commented:
my quick todo list (gotta run for now):
* Revert the things I said I'd revert.
* Update PR title/main description to be up to date with all the changes we made.

BrianPugh commented:
Finally circled back around on this; @kiancross I believe this is ready for another review!

kiancross commented on .github/workflows/continuous-integration.yml:
```diff
@@ -55,113 +55,111 @@ jobs:
+          coveralls --service=github --finish
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          COVERALLS_SERVICE_NUMBER: "${{ github.workflow }}-${{ github.run_number }}"
```
```suggestion
```

kiancross replied:
Think this was left in my accident - just removed as it was breaking CI!

kiancross reviewed (approved):
All looks good, thanks a lot for all your work!

CRD716 commented:
Is it possible to support the latest version of prompt_toolkit yet?

stonebig commented:
maybe it needs prompt-toolkit-3.0.40, not yet released , with this https://github.com/prompt-toolkit/python-prompt-toolkit/pull/1799
